### PR TITLE
feat(discover): Add Validations (Conditions and OrderBy) - WIP

### PR DESCRIPTION
**Pull request by ayesha-omarali**
Validates & tests
* if conditions is a list of lists: `[['error.type', '!=', 'ValidationError']]`
* condition's lists are of length 3
* orderby value is either from fields, aggregate, or is timestamp (for charts)

### CODE REVIEW

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 1]
```diff
@@ -110,6 +110,30 @@ def validate_range(self, attrs, source):
...
+        if attrs.get(source):
+            order_by = attrs.get(source).replace('-', '').replace('+', '')
+            fields = attrs.get('fields')
+            aggregations = attrs.get('aggregations')
+
+            if order_by == 'timestamp' or order_by == 'time':  # timestamp query for charts
+                return attrs
+            elif aggregations and fields:
+                aggregations = set(agg[2] for agg in aggregations)
+                print(aggregations, order_by, any(order_by == agg for agg in aggregations))
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_discover_query.py]: Should this be here?

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 2]
```diff
@@ -110,6 +110,30 @@ def validate_range(self, attrs, source):
...
+                if not any(order_by == agg for agg in aggregations) and not any(order_by == field for field in fields):
+                        raise serializers.ValidationError('Invalid OrderBy - Must be in Fields or Aggregations')
+            elif aggregations:
+                aggregations = set(agg[2] for agg in aggregations)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_discover_query.py]: Will the `2` index always be set or could you hit an IndexError here?

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 3]
```diff
@@ -127,6 +151,14 @@ def validate_conditions(self, attrs, source):
         # Handle error (exception_stacks), stack(exception_frames)
         if attrs.get(source):
             conditions = [self.get_condition(condition) for condition in attrs[source]]
+            invalid_functions = [
+                condition for condition in conditions if not isinstance(
+                    condition, list)]
+            if invalid_functions:
+                raise serializers.ValidationError(
+                    u'Invalid condition function, must be list of lists - {}'.format(
```
[the remark this answers is not in the document]
Comment (ayesha-omarali) [reply src/sentry/api/endpoints/organization_discover_query.py]: thanks, removing it!

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 4]
```diff
@@ -110,6 +110,30 @@ def validate_range(self, attrs, source):
 
         return attrs
 
+    def validate_orderby(self, attrs, source):
+        if attrs.get(source):
+            order_by = attrs.get(source).replace('-', '').replace('+', '')
```
[the remark this answers is not in the document]
Comment (ayesha-omarali) [reply src/sentry/api/endpoints/organization_discover_query.py]: Removing `+`, and I'll add a validation to see if it's in the first position.

File: src/sentry/api/endpoints/organization_discover_query.py [hunk 5]
```diff
@@ -148,6 +180,11 @@ def get_array_field(self, field):
         return re.search(pattern, field)
 
     def get_condition(self, condition):
+        if len(condition) != 3:
+            raise serializers.ValidationError(
+                u'Invalid condition function, must be list containing lists of length 3 - {}'.format(
```
[the remark this answers is not in the document]
Comment (ayesha-omarali) [reply src/sentry/api/endpoints/organization_discover_query.py]: Good point. I think it makes sense to allow `or` - I'll remove it.

**Pull request closed without merging**